crates/proto/src/rr/type_bit_map.rs

```diff
@@ -7,186 +7,276 @@
 
 //! type bit map helper definitions
 
-use alloc::collections::BTreeMap;
+use core::fmt;
+use core::hash::{Hash, Hasher};
+
+use alloc::collections::{
+    BTreeMap, BTreeSet,
+    btree_set::{IntoIter, Iter},
+};
 use alloc::vec::Vec;
 
 use crate::error::*;
-use crate::rr::RecordType;
+use crate::rr::{RecordDataDecodable, RecordType};
 use crate::serialize::binary::*;
 
-enum BitMapReadState {
-    Window,
-    Len {
-        window: u8,
-    },
-    RecordType {
-        window: u8,
-        len: Restrict<u8>,
-        left: Restrict<u8>,
-    },
+/// A compressed bitmap of record types.
```

Comment:
I'd split up this commit. For example:

- Move `BitMapReadState` down
- Replace `Vec<RecordType>` with `TypeBitMaps` (state only)
- Add `original_encoding` to `TypeBitMaps`
- Move `encode_type_bit_maps()` to `RecordDataDecodable` impl
- Move `decode_type_bit_maps()` to `BinEncodable` impl

I also don't love the `TypeBitMaps`, suggest something like `RecordTypeSet` or `TypeSet`. Also suggest that this type should be `pub(crate)` rather than a part of the public API (and avoid making the containing module part of the API -- instead export the type itself from the parent module).